Shield.Communication/Services/Wifi.cs

```diff
@@ -0,0 +1,297 @@
+    OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+    THE SOFTWARE.
+*/
+
+using System;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Windows.ApplicationModel.Core;
+using Windows.Networking;
+using Windows.Networking.Proximity;
+using Windows.Networking.Sockets;
+using Windows.Storage.Streams;
+using Windows.Devices.Bluetooth.Rfcomm;
+using Windows.Devices.Enumeration;
+
+namespace Shield.Communication.Services
+{
+    public class Wifi : ServiceBase
+    {
+        private int port = 1235;
+        private bool beaconIsOn = false;
+        DatagramSocket listener = new DatagramSocket(); 
+
+        public Wifi()
+        {
+            isPollingToSend = true;
+        }
+
+        public override async Task<Connections> GetConnections()
+        {
+            var connections = new Connections();
+            foreach (var client in clients)
+            {
+                if (client.Value != null && client.Value.Source != null)
+                {
+                    var peer = client.Value.Source as RemotePeer;
+                    if (peer != null)
+                    {
+                        if (DateTime.Now.Subtract(peer.Pinged).TotalSeconds > 15)
+                        {
+                            continue;
+                        }
+                    }
+
+                    connections.Add(client.Value);
+                }
+            }
+
+            await Task.FromResult(false);
+            return connections;
+        }
+
+        public override async void ListenForBeacons()
+        {
+            if (!beaconIsOn)
+            {
+                beaconIsOn = true;
+                listener.MessageReceived += ListenerOnMessageReceived;
+                try
+                {
+                    await listener.BindServiceNameAsync(port.ToString());
+                }
+                catch (Exception e)
+                {
+                    // ignore dropped listen sockets
+                }
+            }
+        }
+
+        private async void ListenerOnMessageReceived(DatagramSocket sender, DatagramSocketMessageReceivedEventArgs args)
+        {
+            object outObj;
+            if (CoreApplication.Properties.TryGetValue("remotePeer", out outObj))
+            {
+                EchoMessage((RemotePeer)outObj, args);
+                return;
+            }
+
+            // We do not have an output stream yet so create one.
+            try
+            {
+                IOutputStream outputStream = await listener.GetOutputStreamAsync(args.RemoteAddress, args.RemotePort);
+
+                // It might happen that the OnMessage was invoked more than once before the GetOutputStreamAsync completed.
+                // In this case we will end up with multiple streams - make sure we have just one of it.
+                RemotePeer peer;
+                lock (this)
```

Comment:
Locking on this is generally discouraged. http://stackoverflow.com/questions/251391/why-is-lockthis-bad